apps/gdalalg_dispatcher.h

```diff
@@ -0,0 +1,242 @@
+    {
+        CPLError(CE_Failure, CPLE_AppDefined,
+                 "The Run() method should not be called directly on the \"gdal "
+                 "%s\" program.",
+                 GetName().c_str());
+        return false;
+    }
+};
+
+/************************************************************************/
+/*         GDALDispatcherAlgorithm::ParseCommandLineArguments()         */
+/************************************************************************/
+
+template <class RasterDispatcher, class VectorDispatcher>
+bool GDALDispatcherAlgorithm<RasterDispatcher, VectorDispatcher>::
+    ParseCommandLineArguments(const std::vector<std::string> &args)
+{
+    bool ok;
+    {
+        std::unique_ptr<CPLErrorStateBackuper> oErrorHandler;
+        if (args.size() > 1)
+        {
+            oErrorHandler =
+                std::make_unique<CPLErrorStateBackuper>(CPLQuietErrorHandler);
+        }
+        ok = m_rasterDispatcher->ParseCommandLineArguments(args);
+        CPL_IGNORE_RET_VAL(oErrorHandler);
+    }
+
+    if (m_rasterDispatcher->PropagateSpecialActionTo(this))
+    {
+        return true;
+    }
+
+    if (ok)
+    {
+        auto poDS = m_rasterDispatcher->GetDatasetRef();
+        // cppcheck-suppress knownConditionTrueFalse
+        if (poDS &&
+            (poDS->GetRasterCount() > 0 || poDS->GetMetadata("SUBDATASETS")))
+        {
+            if (poDS->GetLayerCount() != 0)
+            {
+                m_showUsage = false;
+                CPLError(CE_Failure, CPLE_AppDefined,
+                         "'%s' has both raster and vector content. "
+                         "Please use 'gdal raster %s' or 'gdal vector %s'.",
+                         poDS->GetDescription(), GetName().c_str(),
+                         GetName().c_str());
+                return false;
+            }
+
+            m_selectedSubAlg = m_rasterDispatcher.get();
+            std::vector<std::string> callPath(m_callPath);
+            callPath.push_back("raster");
+            m_selectedSubAlg->SetCallPath(callPath);
+
+            return true;
+        }
+    }
+    else if (args.size() <= 1)
+    {
+        return false;
+    }
+
+    auto poDSFromRaster = m_rasterDispatcher->GetDatasetRef();
+    // cppcheck-suppress knownConditionTrueFalse
+    if (poDSFromRaster)
+    {
+        m_vectorDispatcher->SetDataset(poDSFromRaster);
+    }
+
+    std::vector<std::string> argsWithoutInput;
+    bool skipNext = false;
+    for (const auto &arg : args)
+    {
+        if (arg == "-i" || arg == "--input")
+        {
+            skipNext = true;
+        }
+        else if (!skipNext)
+        {
+            if (!STARTS_WITH(arg.c_str(), "--input=") &&
+                !(poDSFromRaster && arg == poDSFromRaster->GetDescription()))
+            {
+                argsWithoutInput.push_back(arg);
+            }
+        }
+        else
+        {
+            skipNext = false;
+        }
+    }
+
+    {
```

Comment:
> what is exactly on the stack?

The CPLErrorStateBackuper.

Reply:
This is a RAII based error handler masking: cf https://github.com/OSGeo/gdal/blob/bb23d74dc14446da751c1763a81909e72e0b70ea/port/cpl_error.h#L284

Reply:
I understood that when I saw the stack-based version. Is there some reason to do it with a unique_ptr earlier?  For me at least it's just easier to follow.

Reply:
oh, I now get your point.
I've rewritten the beginning of the method in a more straightforward way:
```
    bool ok;
    if (args.size() > 1)
    {
        // Silence errors as it might be rather for the vector algorithm
        CPLErrorStateBackuper oErrorHandler(CPLQuietErrorHandler);
        ok = m_rasterDispatcher->ParseCommandLineArguments(args);
    }
    else
    {
        // If there's just a single argument, we don't need to silence errors
        // as this will trigger a legitimate error message about the subcommand.
        ok = m_rasterDispatcher->ParseCommandLineArguments(args);
    }
```